// @sap/eslint-plugin-cds/lib/rules/js/case-sensitive-well-known-events.js
'use strict'

const { RULE_CATEGORIES } = require('../../constants')
const { CdsHandlerRule } = require('./CdsHandlerRule')

const WELL_KNOWN_EVENTS = [
  'CREATE', 'READ', 'UPDATE', 'UPSERT','DELETE',
  'INSERT','SELECT',
  'POST','GET','PUT','PATCH',
  'NEW', 'CANCEL', 'EDIT', 'SAVE'  // draft related
]

class CaseSensitiveWellKnownEvents extends CdsHandlerRule {
  CAPHandlerRegistration(node) {
    const first = node.parent?.arguments?.[0]
    if (!first) return
    const currentEventName = first.value ?? ''
    const currentEventNameUpper = currentEventName.toUpperCase()
    if (WELL_KNOWN_EVENTS.includes(currentEventNameUpper) && currentEventName !== currentEventNameUpper) {
      const quoteType = ['"', '\'', '`'].includes(first.raw?.[0]) ? first.raw?.[0] : '"'
      this.context.report({
        node: first,
        messageId: 'incorrectEventNameCase',
        data: {
          currentEventName: currentEventName,
          properEventName: currentEventNameUpper
        },
        suggest: [{
          desc: 'Change event casing to upper case',
          fix: fixer => fixer.replaceText(first, `${quoteType}${currentEventNameUpper}${quoteType}`)
        }]
      })
    }
  }
}

module.exports = {
  meta: {
    type: 'problem',
    docs: {
      recommended: true,
      category: RULE_CATEGORIES.javascript,
      description: 'Make sure well-known events are used with proper casing.'
    },
    fixable: 'code',
    schema: [],
    messages: {
      incorrectEventNameCase: 'Found an event registration for event "{{currentEventName}}", which is likely supposed to be "{{properEventName}}".'
    },
    hasSuggestions: true
  },
  create: context => new CaseSensitiveWellKnownEvents(context).asESLintVisitor()
}
